test/contrib/test_funsor.py

```diff
@@ -0,0 +1,190 @@
+
+from jax import random
+import jax.numpy as jnp
+
+from funsor import Tensor, bint, reals
+import numpyro
+from numpyro.contrib.funsor.enum_messenger import NamedMessenger, markov, to_data, to_funsor
+import numpyro.distributions as dist
+from numpyro.infer import MCMC, NUTS
+
+
+def test_gaussian_mixture_model():
+    K, N = 3, 1000
+
+    def gmm(data):
+        mix_proportions = numpyro.sample("phi", dist.Dirichlet(jnp.ones(K)))
+        with numpyro.plate("num_clusters", K, dim=-1):
+            cluster_means = numpyro.sample("cluster_means", dist.Normal(jnp.arange(K), 1.))
+        with numpyro.plate("data", data.shape[0], dim=-1):
+            assignments = numpyro.sample("assignments", dist.Categorical(mix_proportions))
+            numpyro.sample("obs", dist.Normal(cluster_means[assignments], 1.), obs=data)
+
+    true_cluster_means = jnp.array([1., 5., 10.])
+    true_mix_proportions = jnp.array([0.1, 0.3, 0.6])
+    cluster_assignments = dist.Categorical(true_mix_proportions).sample(random.PRNGKey(0), (N,))
+    data = dist.Normal(true_cluster_means[cluster_assignments], 1.0).sample(random.PRNGKey(1))
+
+    nuts_kernel = NUTS(gmm)
+    mcmc = MCMC(nuts_kernel, num_warmup=500, num_samples=500)
+    mcmc.run(random.PRNGKey(2), data)
+    samples = mcmc.get_samples()
+    assert_allclose(samples["phi"].mean(0).sort(), true_mix_proportions, atol=0.05)
+    assert_allclose(samples["cluster_means"].mean(0).sort(), true_cluster_means, atol=0.2)
+
+
+def test_bernoulli_latent_model():
+    def model(data):
+        y_prob = numpyro.sample("y_prob", dist.Beta(1., 1.))
+        with numpyro.plate("data", data.shape[0]):
+            y = numpyro.sample("y", dist.Bernoulli(y_prob))
+            z = numpyro.sample("z", dist.Bernoulli(0.65 * y + 0.1))
+            numpyro.sample("obs", dist.Normal(2. * z, 1.), obs=data)
+
+    N = 2000
+    y_prob = 0.3
+    y = dist.Bernoulli(y_prob).sample(random.PRNGKey(0), (N,))
+    z = dist.Bernoulli(0.65 * y + 0.1).sample(random.PRNGKey(1))
+    data = dist.Normal(2. * z, 1.0).sample(random.PRNGKey(2))
+
+    nuts_kernel = NUTS(model)
+    mcmc = MCMC(nuts_kernel, num_warmup=500, num_samples=500)
+    mcmc.run(random.PRNGKey(3), data)
+    samples = mcmc.get_samples()
+    assert_allclose(samples["y_prob"].mean(0), y_prob, atol=0.05)
+
+
+def test_change_point():
+    def model(count_data):
+        n_count_data = count_data.shape[0]
+        alpha = 1 / jnp.mean(count_data)
+        lambda_1 = numpyro.sample('lambda_1', dist.Exponential(alpha))
+        lambda_2 = numpyro.sample('lambda_2', dist.Exponential(alpha))
+        # this is the same as DiscreteUniform(0, 69)
+        tau = numpyro.sample('tau', dist.Categorical(logits=jnp.zeros(70)))
+        idx = jnp.arange(n_count_data)
+        lambda_ = jnp.where(tau > idx, lambda_1, lambda_2)
+        with numpyro.plate("data", n_count_data):
+            numpyro.sample('obs', dist.Poisson(lambda_), obs=count_data)
```

Comment:
@eb8680 Could you help me debug this test? I think currently, enum only works when discrete sites are inside a `plate` statement. :(

Reply:
Just pushed a fix - `enum` wasn't properly passing calls to `to_data` to its parent handler.

Reply:
Awesome! Thanks a lot, Eli!